# Identification-of-differentially-expressed-genes-related-to-HCV-induced-HCC
Chronic hepatitis C virus (HCV) infection is a major cause of advanced hepatic fibrosis and cirrhosis, with significantly increased risk for development of hepatocellular carcinoma (HCC); the fifth most common cancer and second most cause of cancer-related death worldwide. HCC development due to HCV is a stepwise process spanning over 20 to 30 years. HCV carcinogenesis is mediated by two major factors; viral-induced factors and host-induced immunologic response. To understand the alteration from benign to malignant, we analyzed the gene expression in liver tissues at different phases, including normal, cirrhosis, and different HCC stages. We searched for datasets in major transcriptome profiling data bases and selected Mas VR et al. â€˜s micro array dataset for our analysis. Differentially expressed genes (DEGs) were identified using DESeq2 after adjustment for false-discovery rate (Adjusted P-value < 0.05) and more than 1 log Fold change (LFC). Gene Set Enrichment Analysis (GSEA) was also performed to identify Go-terms and biological pathways involved in HCV-induced HCC. Prediction of miRNA that targets the unique DEGs was performed using MiRWalk. Prediction of potential small-molecule signatures was executed by L1000CDS2. We identified 69 differential expressed genes that may contribute in the induction of HCV to HCC. This study also found 22 highly enriched GO terms and 79 miRNAs that target the DEGs. We identified hsa-mir-21, which expressed in peripheral blood may be considered as a good predictor bio marker for HCV related carcinogenesis. It was predicted also that small-molecules Withaferin-a, Arachidonyl trifluoro-methyl ketone, Perhexiline, Penfluridol and Bufalin can potentially reverse gene expression in HCV-induced HCC and could be used for monitoring the treatment of HCV-carcinogenesis.
